src/adr/0018-polyglot-strategy.md

```diff
@@ -0,0 +1,158 @@
+# 16. Polyglot codebase
+
+Date: 2020-02-03
+
+## Status
+
+Proposed
+
+## Context & Problem Statement
+
+James is written in Java for a very long time. In recent years, Java modernized a lot after a decade of slow progress.
+
+However, in the meantime, most software relying on the JVM started supporting alternative JVM languages to keep being relevant.
+
+It includes Groovy, Clojure, Scala and more recently Kotlin.
+
+Not being open to those alternative languages can be a problem for James adoption.
+
+## Decision drivers
+
+Nowadays, libraries and framework targeting the JVM are expected to support usage of one or several of these alternative languages.
+
+James being not only a mail server but also a development framework needs to reach those expectations.
+
+At the same time, more and more developers and languages adopt Function Programming (FP) idioms to solve their problems.
+
+## Considered options
+
+1. Let the users figure out how to make polyglot setups
+2. Document the usage of polyglot mailets for some popular languages
+3. Document the usage of polyglot components for some popular languages
+4. Actually implement some mailets in some popular languages
+5. Actually implement some components in some popular languages
+
+About Languages:
+
+I. Clojure
+II. Groovy
+III. Kotlin
+IV. Scala
+
+## Decision
+
+We decided on options 4, 5 and IV.
+
+That means we need to write some mailets in Scala and demonstrate how it's done and then used in a running server.
+
+It also means writing and/or refactoring some server components in Scala, starting where it's the most relevant.
+
+### Positive Consequences 
+
+* Modernize parts of James code
+* Leverage Scala richer FP ecosystem and language to overcome Java limitations on that topic
+* Should attract people that would not like Java
+
+### Negative Consequences 
+
+* Adds even more knowledge requirements to contribute to James
```

Comment:
Defines 'contribute'. IMO you ease or not impact small contributiuons (as you accept one more language, and still accept java only contributions).

However I agree the overall system understanding is harder (as you have to also master to some extents scala)

Reply:
Do you really want me to define contribute? I mean, I share your analysis but I don't see how refining 'contribute' would make this point different. Opinion?

Reply:
Well, my point is more to dplit this argument in 2: one positive, one negative